snapcraft/internal/meta.py

```diff
@@ -134,15 +136,21 @@ def setup_assets(self):
             file_utils.link_or_copy(
                 'gadget.yaml', os.path.join(self.meta_dir, 'gadget.yaml'))
 
-    def _record_manifest(self):
-        manifest_dir = os.path.join(self._prime_dir, 'snap')
-        manifest_file_path = os.path.join(manifest_dir, 'manifest.yaml')
+    def _record_manifest_and_source_snapcraft_yaml(self):
+        prime_snap_dir = os.path.join(self._prime_dir, 'snap')
+        recorded_snapcraft_yaml_path = os.path.join(
+            prime_snap_dir, 'snapcraft.yaml')
+        if os.path.isfile(recorded_snapcraft_yaml_path):
+            os.unlink(recorded_snapcraft_yaml_path)
+        manifest_file_path = os.path.join(prime_snap_dir, 'manifest.yaml')
         if os.path.isfile(manifest_file_path):
             os.unlink(manifest_file_path)
 
         # FIXME hide this functionality behind a feature flag for now
         if os.environ.get('SNAPCRAFT_BUILD_INFO'):
-            os.makedirs(manifest_dir, exist_ok=True)
+            os.makedirs(prime_snap_dir, exist_ok=True)
+            shutil.copy2(
+                self._snapcraft_yaml_path, recorded_snapcraft_yaml_path)
```

Comment:
There were a few issues here.
Previously, we were recording the snapcraft.yaml even when the flag is not set, which is leaking information. But that happened only when the yaml was in snap/snapcraft.yaml. The other two locations were ignored.
I saw four options to solve this. One, find the snapcraft.yaml in meta. That's basically, do the same thing we do in load, once at the start and once at the end. The second, save in the project options the path of the loaded yaml. Third, make public the snapcraft_yaml_path attribute that we have in config, and pass it to the function that creates meta at the end. Fourth, pass to meta the whole config, not just the loaded data.
I hated all those options, so I went with the easiest, which was #3.